programs/vote/src/vote_state/mod.rs

```diff
@@ -96,6 +96,48 @@ impl Lockout {
     }
 }
 
+#[derive(Serialize, Default, Deserialize, Debug, PartialEq, Eq, Clone, AbiExample)]
+pub struct VoteStateUpdate {
+    /// The proposed tower
+    pub lockouts: VecDeque<Lockout>,
+    /// The proposed root
+    pub root: Option<Slot>,
+    /// signature of the bank's state at the last slot
+    pub hash: Hash,
+    /// processing timestamp of last slot
+    pub timestamp: Option<UnixTimestamp>,
+}
+
+impl VoteStateUpdate {
+    pub fn new(lockouts: VecDeque<Lockout>, root: Option<Slot>, hash: Hash) -> Self {
+        Self {
+            lockouts,
+            root,
+            hash,
+            timestamp: None,
+        }
+    }
+
+    pub fn last_voted_slot(&self) -> Option<Slot> {
+        self.lockouts.back().copied().map(|lockout| lockout.slot)
+    }
+
+    pub fn last_voted_slot_hash(&self) -> Option<(Slot, Hash)> {
+        self.lockouts
+            .back()
+            .copied()
+            .map(|lockout| (lockout.slot, self.hash))
+    }
+
+    pub fn get_legacy_vote(&self) -> Vote {
+        Vote {
+            slots: self.lockouts.iter().map(|&lockout| lockout.slot).collect(),
```

Comment:
If we implemented a common trait `VoteTransaction` (or a better name) for `VoteStateUpdate` and `Vote`, and passed that trait to `check_slots_are_valid()`, we could avoid this clone.

Reply:
yeah I agree, seems wasteful right now, i'll do that